File: src/test/java/io/tolgee/controllers/RepositoryControllerTest.java
package io.tolgee.controllers;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.tolgee.dtos.request.AbstractRepositoryDTO;
import io.tolgee.dtos.request.CreateRepositoryDTO;
import io.tolgee.dtos.request.EditRepositoryDTO;
import io.tolgee.dtos.request.LanguageDTO;
import io.tolgee.dtos.response.RepositoryDTO;
import io.tolgee.helpers.JsonHelper;
import io.tolgee.model.Language;
import io.tolgee.model.Repository;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.annotation.Rollback;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.transaction.annotation.Transactional;
import org.testng.annotations.Test;

import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Optional;
import java.util.Set;

import static io.tolgee.controllers.LoggedRequestFactory.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class RepositoryControllerTest extends SignedInControllerTest {
    private final LanguageDTO languageDTO = new LanguageDTO(null, "English", "en");

    @Test
    void createRepository() throws Exception {
        dbPopulator.createBase("test");

        testCreateValidationSize();

        testCreateValidationUniqueness();

        testCreateCorrectRequest();
    }

    private void testCreateCorrectRequest() throws Exception {
        CreateRepositoryDTO request = new CreateRepositoryDTO("aaa", Collections.singleton(languageDTO));

        MvcResult mvcResult = mvc.perform(
                loggedPost("/api/repositories")
                        .contentType(MediaType.APPLICATION_JSON).content(
                        JsonHelper.asJsonString(request)))
                .andExpect(status().isOk())
                .andReturn();

        Optional<Repository> aa = repositoryService.findByName("aaa", userAccount);

        assertThat(aa).isPresent();

        Repository repository = aa.orElse(null);

        assertThat(repository.getLanguages()).isNotEmpty();

        Language language = repository.getLanguages().stream().findFirst().orElse(null);

        assertThat(language).isNotNull();

        assertThat(language.getAbbreviation()).isEqualTo("en");

        assertThat(language.getName()).isEqualTo("English");
    }

    private void testCreateValidationSize() throws Exception {
        CreateRepositoryDTO request = new CreateRepositoryDTO("aa", Collections.singleton(languageDTO));

        //test validation
        MvcResult mvcResult = mvc.perform(
                loggedPost("/api/repositories")
                        .contentType(MediaType.APPLICATION_JSON).content(
                        JsonHelper.asJsonString(request)))
                .andExpect(status().isBadRequest())
                .andReturn();

        assertThat(mvcResult.getResponse().getContentAsString()).contains("name");
        assertThat(mvcResult.getResponse().getContentAsString()).contains("STANDARD_VALIDATION");
    }

    private void testCreateValidationUniqueness() throws Exception {
        CreateRepositoryDTO request = new CreateRepositoryDTO("test", Collections.singleton(languageDTO));

        //test validation
        MvcResult mvcResult = mvc.perform(
                loggedPost("/api/repositories")
                        .contentType(MediaType.APPLICATION_JSON).content(
                        JsonHelper.asJsonString(request)))
                .andExpect(status().isBadRequest())
                .andReturn();

        assertThat(mvcResult.getResponse().getContentAsString())
                .isEqualTo("{\"STANDARD_VALIDATION\":{\"name\":\"NAME_EXISTS\"}}");
    }


    @Test
    void editRepository() throws Exception {
        Repository test = dbPopulator.createBase(generateUniqueString());
        Repository test2 = dbPopulator.createBase(generateUniqueString());

        testEditValidationUniqueness(test, test2.getName());
        testEditCorrectRequest(test);
    }

    private void testEditCorrectRequest(Repository test) throws Exception {
        AbstractRepositoryDTO request = new EditRepositoryDTO(test.getId(), "new test");

        MvcResult mvcResult = mvc.perform(
                loggedPost("/api/repositories/edit")
                        .contentType(MediaType.APPLICATION_JSON).content(
                        JsonHelper.asJsonString(request)))
                .andExpect(status().isOk()).andReturn();

        ObjectMapper mapper = new ObjectMapper();
        RepositoryDTO response = mapper.readValue(mvcResult.getResponse().getContentAsString(), RepositoryDTO.class);

        assertThat(response.getName()).isEqualTo("new test");
        assertThat(response.getId()).isEqualTo(test.getId());

        Optional<Repository> found = repositoryService.findByName("new test", userAccount);

        assertThat(found).isPresent();
    }

    private void testEditValidationUniqueness(Repository repository, String nonUniqueName) throws Exception {
        EditRepositoryDTO request = new EditRepositoryDTO(repository.getId(), nonUniqueName);

        //test validation
        MvcResult mvcResult = mvc.perform(
                loggedPost("/api/repositories/edit")
                        .contentType(MediaType.APPLICATION_JSON).content(
                        JsonHelper.asJsonString(request)))
                .andExpect(status().isBadRequest())
                .andReturn();

        assertThat(mvcResult.getResponse().getContentAsString())
                .isEqualTo("{\"STANDARD_VALIDATION\":{\"name\":\"NAME_EXISTS\"}}");
    }

    @Test
    @Rollback
    void deleteRepository() throws Exception {
        Repository test = dbPopulator.createBase(generateUniqueString());

        MvcResult mvcResult = mvc.perform(
                loggedDelete("/api/repositories/" + test.getId())
                        .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andReturn();

        Optional<Repository> repository = repositoryService.findById(test.getId());

        assertThat(repository).isEmpty();
    }


    @Test
    void findAll() throws Exception {
        LinkedHashSet<String> repos = new LinkedHashSet<>();
        for (int i = 0; i < 3; i++) {
            repos.add(dbPopulator.createBase(generateUniqueString()).getName());
        }

        MvcResult mvcResult = mvc.perform(
                loggedGet("/api/repositories/")
                        .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andReturn();

        @SuppressWarnings("unchecked") Set<RepositoryDTO> set = mapResponse(mvcResult, Set.class, RepositoryDTO.class);
        assertThat(set).extracting("name").containsAll(repos);
    }
}

File: src/test/java/io/tolgee/controllers/RepositoryControllerTest.kt
package io.tolgee.controllers

import com.fasterxml.jackson.databind.ObjectMapper
import io.tolgee.assertions.Assertions.assertThat
import io.tolgee.dtos.request.CreateRepositoryDTO
import io.tolgee.dtos.request.EditRepositoryDTO
import io.tolgee.dtos.request.LanguageDTO
import io.tolgee.dtos.response.RepositoryDTO
import io.tolgee.fixtures.LoggedRequestFactory.loggedDelete
import io.tolgee.fixtures.LoggedRequestFactory.loggedGet
import io.tolgee.fixtures.LoggedRequestFactory.loggedPost
import io.tolgee.fixtures.andIsOk
import io.tolgee.fixtures.generateUniqueString
import io.tolgee.fixtures.mapResponseTo
import io.tolgee.helpers.JsonHelper
import io.tolgee.model.Repository
import org.assertj.core.api.Assertions
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.annotation.Rollback
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.testng.annotations.Test
import java.util.*

@SpringBootTest
@AutoConfigureMockMvc
class RepositoryControllerTest : SignedInControllerTest() {
    private val languageDTO = LanguageDTO(null, "English", "en")

    @Test
    fun createRepository() {
        dbPopulator.createBase("test")
        testCreateValidationSize()
        testCreateCorrectRequest()
    }

    @Test
    fun createRepositoryOrganization() {
        val userAccount = dbPopulator.createUser("testuser")
        val organization = dbPopulator.createOrganization("Test Organization", userAccount)
        logAsUser("testuser", initialPasswordManager.initialPassword)
        val request = CreateRepositoryDTO("aaa", setOf(languageDTO), organizationId = organization.id)
        val result = performAuthPost("/api/repositories", request).andIsOk.andReturn().mapResponseTo<RepositoryDTO>()
        repositoryService.get(result.id!!).get().let {
            assertThat(it.organizationOwner?.id).isEqualTo(organization.id)
        }
    }


    private fun testCreateCorrectRequest() {
        val request = CreateRepositoryDTO("aaa", setOf(languageDTO))
        val mvcResult = mvc.perform(
                loggedPost("/api/repositories")
                        .contentType(MediaType.APPLICATION_JSON).content(
                                JsonHelper.asJsonString(request)))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()
        val repositoryDto = repositoryService.findAllPermitted(userAccount!!).find { it.name == "aaa" }
        Assertions.assertThat(repositoryDto).isNotNull
        val repository = repositoryService.get(repositoryDto!!.id!!).get()
        Assertions.assertThat(repository.languages).isNotEmpty
        val language = repository.languages.stream().findFirst().orElse(null)
        Assertions.assertThat(language).isNotNull
        Assertions.assertThat(language.abbreviation).isEqualTo("en")
        Assertions.assertThat(language.name).isEqualTo("English")
    }

    private fun testCreateValidationSize() {
        val request = CreateRepositoryDTO("aa", setOf(languageDTO))

        //test validation
        val mvcResult = mvc.perform(
                loggedPost("/api/repositories")
                        .contentType(MediaType.APPLICATION_JSON).content(
                                JsonHelper.asJsonString(request)))
                .andExpect(MockMvcResultMatchers.status().isBadRequest)
                .andReturn()
        Assertions.assertThat(mvcResult.response.contentAsString).contains("name")
        Assertions.assertThat(mvcResult.response.contentAsString).contains("STANDARD_VALIDATION")
    }

    @Test
    fun editRepository() {
        val test = dbPopulator.createBase(generateUniqueString())
        testEditCorrectRequest(test)
    }

    private fun testEditCorrectRequest(test: Repository) {
        val request = EditRepositoryDTO(test.id, "new test")
        val mvcResult = mvc.perform(
                loggedPost("/api/repositories/edit")
                        .contentType(MediaType.APPLICATION_JSON).content(
                                JsonHelper.asJsonString(request)))
                .andExpect(MockMvcResultMatchers.status().isOk).andReturn()
        val mapper = ObjectMapper()
        val response = mapper.readValue(mvcResult.response.contentAsString, RepositoryDTO::class.java)
        Assertions.assertThat(response.name).isEqualTo("new test")
        Assertions.assertThat(response.id).isEqualTo(test.id)
        val found = repositoryService.findAllPermitted(userAccount!!).find { it.name == "new test" }
        Assertions.assertThat(found).isNotNull
    }

    @Test
    @Rollback
    fun deleteRepository() {
        val base = dbPopulator.createBase(generateUniqueString())
        mvc.perform(
                loggedDelete("/api/repositories/${base.id}")
                        .contentType(MediaType.APPLICATION_JSON))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()
        val repository = repositoryService.get(base.id)
        Assertions.assertThat(repository).isEmpty
    }

    @Test
    fun findAll() {
        val repos = LinkedHashSet<String?>()
        for (i in 0..2) {
            repos.add(dbPopulator.createBase(generateUniqueString()).name)
        }
        val mvcResult = mvc.perform(
                loggedGet("/api/repositories/")
                        .contentType(MediaType.APPLICATION_JSON))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()
        val set: Set<RepositoryDTO> = mvcResult.mapResponseTo()
        Assertions.assertThat(set).extracting("name").containsAll(repos)
    }
}
